pkg/agent/openflow/client_test.go

```diff
@@ -1123,6 +1126,73 @@ func Test_client_InstallServiceGroup(t *testing.T) {
 	}
 }
 
+func getServiceBucketMaxSize(t *testing.T, ctrl *gomock.Controller) uint16 {
+	fakeOfTable := ovsoftest.NewMockTable(ctrl)
+	fakeNodeIPChecker := nodeiptest.NewFakeNodeIPChecker("2001::2")
+	ServiceLBTable.ofTable = fakeOfTable
+	defer func() {
+		ServiceLBTable.ofTable = nil
+	}()
+	fs := &featureService{
+		groupCache:    sync.Map{},
+		bridge:        binding.NewOFBridge(bridgeName, ""),
+		nodeIPChecker: fakeNodeIPChecker,
+	}
+	// To create a single-bucket group with maximum size, use a remote non-hostNetwork IPv6 Endpoint that has all
+	// available actions.
+	endpoint := proxy.NewBaseEndpointInfo("2001::1", "node1", "", 80, false, true, false, false, nil)
+	fakeOfTable.EXPECT().GetID().Return(uint8(1)).Times(1)
+	group := fs.serviceEndpointGroup(binding.GroupIDType(100), true, endpoint)
+
+	messages, err := group.GetBundleMessages(binding.AddMessage)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(messages))
+	groupMod := messages[0].GetMessage().(*openflow15.GroupMod)
+	require.Equal(t, 1, len(groupMod.Buckets))
+	// Since this is a single-bucket group, the `BucketArrayLen` is the size of the bucket.
+	t.Logf("The maxinum size of a Service group bucket is %d bytes", groupMod.BucketArrayLen)
+	return groupMod.BucketArrayLen
+}
+
+func getMulticastBucketMaxSize(t *testing.T, ctrl *gomock.Controller) uint16 {
+	fs := &featureMulticast{
+		groupCache: sync.Map{},
+		bridge:     binding.NewOFBridge(bridgeName, ""),
+	}
+	fakeOfTable := ovsoftest.NewMockTable(ctrl)
+	MulticastOutputTable.ofTable = fakeOfTable
+	defer func() {
+		MulticastOutputTable.ofTable = nil
+	}()
+
+	// Create a single-bucket group with maximum size.
+	fakeOfTable.EXPECT().GetID().Return(uint8(1)).Times(1)
+	group := fs.multicastReceiversGroup(binding.GroupIDType(100), 0, nil, []net.IP{net.ParseIP("192.168.77.100")})
+
+	messages, err := group.GetBundleMessages(binding.AddMessage)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(messages))
+	groupMod := messages[0].GetMessage().(*openflow15.GroupMod)
+	require.Equal(t, 1, len(groupMod.Buckets))
+	// Since this is a single-bucket group, the `BucketArrayLen` is the size of the bucket.
+	t.Logf("The maxinum size of a Multicast group bucket is %d bytes", groupMod.BucketArrayLen)
+	return groupMod.BucketArrayLen
+}
+
+func TestMaxBucketsPerMessage(t *testing.T) {
+	ctrl := gomock.NewController(t)
+
+	groupModHeaderSize := uint16(unsafe.Sizeof(openflow15.GroupMod{}) - unsafe.Sizeof(openflow15.GroupMod{}.Buckets) - unsafe.Sizeof(openflow15.GroupMod{}.Properties))
```

Comment:
you should call `openflow15.GroupMod{}.Len()` instead
even though you are removing the size of the slice descriptors, there is no guarantee that `groupModHeaderSize` will match the message wire size. `unsafe.Sizeof` is platform-dependent, and it is possible for padding to be introduced between struct fields for alignment purposes

Reply:
Make sense, so detailed explanation!